<a href="https://colab.research.google.com/github/kimhj010502/ESAA_OB/blob/main/WEEK12_HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <b> [파이썬 머신러닝 완벽 가이드] </b> <br>
## Chapter 09 추천 시스템

## 9.5 콘텐츠 기반 필터링 실습 - TMDB 5000 영화 데이터 세트 <br>
### 장르 속성을 이용한 영화 콘텐츠 기반 필터링 <br>
- 콘텐츠 기반 필터링은 사용자가 특정 영화를 감상하고 그 영화를 좋아했다면 그 영화와 비슷한 특성/속성, 구성 요소 등을 가진 다른 영화를 추천하는 것임
- 영화 간의 유사성을 판단하는 기준이 영화를 구성하는 다양한 콘텐츠 (장르, 감독, 배우, 평점, 키워드, 영화 설명)를 기반으로 하는 방식임



### 데이터 로딩 및 가공

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [16]:
import pandas as pd
import numpy as np

movies = pd.read_csv('/content/drive/My Drive/ESAA/OB/Data/tmdb_5000_movies.csv')
print(movies.shape)
movies.head(1)

(4803, 20)


budget  \
0  237000000   

                                                                                                genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                                                                              keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "sp...   

  original_language original_title  \
0                en         Avatar   

                                                                                              overview  \
0  In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, ...   

   popularity  \
0  150.437577   

                                                                                  production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289}, {"name": "Twentieth Century Fox Film Corporatio...   

                                                                                  production_countries  \
0  [{"iso_3166_1": "US", "name": "United States of America"}, {"iso_3166_1": "GB", "name": "United ...   

  release_date     revenue  runtime  \
0   2009-12-10  2787965087    162.0   

                                                                        spoken_languages  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso_639_1": "es", "name": "Espa\u00f1ol"}]   

     status                      tagline   title  vote_average  vote_count  
0  Released  Enter the World of Pandora.  Avatar           7.2       11800

- id
- title: 영화 제목
- genres: 영화가 속한 여러 가지 장르
- vote_average: 평균 평점
- vote_count: 평점 투표 수
- popularity: 영화의 인기
- keywords: 영화를 설명하는 주요 키워드 문구
- overview: 영화에 대한 개요 설명

In [17]:
movies_df = movies[['id','title','genres','vote_average','vote_count','popularity','keywords','overview']]

In [18]:
movies_df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                                                                                genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {...   
1       [{"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 28, "name": "Action"}]   
2         [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 80, "name": "Crime"}]   
3  [{"id": 28, "name": "Action"}, {"id": 80, "name": "Crime"}, {"id": 18, "name": "Drama"}, {"id": ...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 878, "name": "Science Fic...   

   vote_average  vote_count  popularity  \
0           7.2       11800  150.437577   
1           6.9        4500  139.082615   
2           6.3        4466  107.376788   
3           7.6        9106  112.312950   
4           6.1        2124   43.926995   

                                                                                              keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "sp...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "name": "drug abuse"}, {"id": 911, "name": "exotic is...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name": "based on novel"}, {"id": 4289, "name": "secret...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853, "name": "crime fighter"}, {"id": 949, "name": "te...   
4  [{"id": 818, "name": "based on novel"}, {"id": 839, "name": "mars"}, {"id": 1456, "name": "medal...   

                                                                                              overview  
0  In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, ...  
1  Captain Barbossa, long believed to be dead, has come back to life and is headed to the edge of t...  
2  A cryptic message from Bond’s past sends him on a trail to uncover a sinister organization. Whil...  
3  Following the death of District Attorney Harvey Dent, Batman assumes responsibility for Dent's c...  
4  John Carter is a war-weary, former military captain who's inexplicably transported to the myster...

- genres와 keywords 칼럼은 내부에 여러 개의 딕셔너리(dict)가 있는 형태의 문자열로 표기되어 있음

In [19]:
pd.set_option('max_colwidth', 100)
movies_df[['genres','keywords']][:1]

genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {...   

                                                                                              keywords  
0  [{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "sp...

In [20]:
#문자열을 분해해서 리스트 객체로 추출
from ast import literal_eval
movies_df['genres'] = movies_df['genres'].apply(literal_eval) #문자열을 객체로 변환
movies_df['keywords'] = movies_df['keywords'].apply(literal_eval)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [21]:
movies_df[['genres','keywords']][:1]

genres  \
0  [{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {...   

                                                                                              keywords  
0  [{'id': 1463, 'name': 'culture clash'}, {'id': 2964, 'name': 'future'}, {'id': 3386, 'name': 'sp...

In [22]:
movies_df['genres'] = movies_df['genres'].apply(lambda x: [y['name'] for y in x]) #리스트 내의 여러 개의 딕셔너리의 'name' 키에 해당하는 값을 찾아 리스트 객체로 변환
movies_df['keywords'] = movies_df['keywords'].apply(lambda x: [y['name'] for y in x])
movies_df[['genres','keywords']][:1]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                                                                              keywords  
0  [culture clash, future, space war, space colony, society, space travel, futuristic, romance, spa...

### 장르 콘텐츠 유사도 측정 <br>
- 문자열로 변환된 genres 칼럼을 Count 기반으로 피처 벡터화 변환함
- genres 문자열을 피처 벡터화 행렬로 변환한 데이터 세트를 코사인 유사도를 통해 비교함. 이를 위해 데이터 세트의 레코드별로 타 레코드와 장르에서 코사인 유사도 값을 가지는 객체를 생성함
-장르 유사도가 높은 영화 중에 평점이 높은 순으로 영화를 추천함

In [23]:
from sklearn.feature_extraction.text import CountVectorizer

#CountVectorizer를 적용하기 위해 공백 문자로 word 단위가 구분되는 문자열로 변환
movies_df['genre_literal'] = movies_df['genres'].apply(lambda x: (' ').join(x))
count_vect = CountVectorizer(min_df=0, ngram_range=(1,2))
genre_mat = count_vect.fit_transform(movies_df['genre_literal'])
print(genre_mat.shape)

(4803, 276)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [26]:
print(genre_mat)

  (0, 0)	1
  (0, 16)	1
  (0, 124)	1
  (0, 232)	1
  (0, 138)	1
  (0, 1)	1
  (0, 24)	1
  (0, 135)	1
  (0, 233)	1
  (1, 0)	1
  (1, 16)	1
  (1, 124)	1
  (1, 24)	1
  (1, 125)	1
  (2, 0)	1
  (2, 16)	1
  (2, 1)	1
  (2, 64)	1
  (2, 20)	1
  (3, 0)	1
  (3, 64)	1
  (3, 90)	1
  (3, 234)	1
  (3, 4)	1
  (3, 68)	1
  :	:
  (4796, 90)	1
  (4796, 234)	1
  (4796, 106)	1
  (4796, 144)	1
  (4797, 234)	1
  (4797, 153)	1
  (4797, 157)	1
  (4798, 0)	1
  (4798, 64)	1
  (4798, 234)	1
  (4798, 4)	1
  (4798, 78)	1
  (4799, 44)	1
  (4799, 214)	1
  (4799, 58)	1
  (4800, 90)	1
  (4800, 44)	1
  (4800, 214)	1
  (4800, 104)	1
  (4800, 50)	1
  (4800, 250)	1
  (4800, 182)	1
  (4800, 251)	1
  (4800, 229)	1
  (4802, 80)	1


<img src='https://ifh.cc/g/AHxOsH.jpg' border='0' width='500'>

In [28]:
from sklearn.metrics.pairwise import cosine_similarity

genre_sim = cosine_similarity(genre_mat, genre_mat)
print(genre_sim.shape)
print(genre_sim[:1])

(4803, 4803)
[[1.         0.59628479 0.4472136  ... 0.         0.         0.        ]]


In [32]:
genre_sim_sorted_ind = genre_sim.argsort()[:,::-1] #높은 순으로 정렬된 비교 행 위치 인덱스 값
print(genre_sim_sorted_ind[:1])

[[   0 3494  813 ... 3038 3037 2401]]


- 0번 레코드의 경우 자신인 0번 레코드를 제외하면 3494번 레코드가 가장 유사도 높고, 그 다음이 813번 레코드이며, 가장 유사도가 낮은 레코드는 2401번 레코드라는 뜻임


### 장르 콘텐츠 필터링을 이용한 영화 추천 <br>
- `find_sim_movie()`는 기반 데이터인 movies_df DataFrame, 레코드별 장르 코사인 유사도 인덱스를 가지고 있는 genre_sim_sorted_ind, 고객이 선정한 추천 기준이 되는 영화 제목, 추천할 영화 건수를 입력하면 추천 영화 정보를 가지는 DataFrame을 반환함


In [33]:
def find_sim_movie(df, sorted_ind, title_name, top_n=10):
    #인자로 입력된 movies_df DataFrame에서 'title' 칼럼이 입력된 title_name 값인 DataFrame 추출
    title_movie = df[df['title'] == title_name]

    #title_name을 가진 DataFrame의 index 객체를 ndarray로 반환하고
    #sorted_ind 인자로 입력된 genre_sim_sorted_ind 객체에서 유사도 순으로 top_n개의 index 추출
    title_index = title_movie.index.values
    similar_indexes = sorted_ind[title_index, :(top_n)]

    #추출된 top_n index 출력, top_n index는 2차원 데이터임
    #dataframe에서 index로 사용하기 위해서 1차원 array로 변경
    print(similar_indexes)
    similar_indexes = similar_indexes.reshape(-1)

    return df.iloc[similar_indexes]

In [35]:
similar_movies = find_sim_movie(df=movies, sorted_ind=genre_sim_sorted_ind, title_name='The Godfather', top_n=10)
similar_movies[['title','vote_average']]

[[2731 1243 3636 1946 2640 4065 1847 4217  883 3866]]


title  vote_average
2731                          The Godfather: Part II           8.3
1243                                    Mean Streets           7.2
3636                                   Light Sleeper           5.7
1946  The Bad Lieutenant: Port of Call - New Orleans           6.0
2640         Things to Do in Denver When You're Dead           6.7
4065                                      Mi America           0.0
1847                                      GoodFellas           8.2
4217                                            Kids           6.8
883                              Catch Me If You Can           7.7
3866                                     City of God           8.1

- 평점이 낮은 영화도 추천되기 때문에 좀 더 많은 후보군을 선정한 뒤에 영화의 평점에 따라 필터링해서 최종 추천하는 방식으로 변경
- 영화의 평점 정보인 'vote_average' 값을 이용하는데, 소수의 관객이 특정 영화에 매우 높은 평점을 부여해 왜곡된 데이터를 가지고 있음

In [36]:
movies_df[['title','vote_average','vote_count']].sort_values('vote_average', ascending=False)[:10]

title  vote_average  vote_count
3519          Stiff Upper Lips          10.0           1
4247     Me You and Five Bucks          10.0           2
4045     Dancer, Texas Pop. 81          10.0           1
4662            Little Big Top          10.0           1
3992                 Sardaarji           9.5           2
2386            One Man's Hero           9.3           2
2970        There Goes My Baby           8.5           2
1881  The Shawshank Redemption           8.5        8205
2796     The Prisoner of Zenda           8.4          11
3337             The Godfather           8.4        5893

- 유명한 영화 평점 사이트인 IMDB에서는 평가 횟수에 대한 가중치가 부여된 평점 (Weighted Rating) 방식을 사용함 <br><br>
$$ 가중 \space 평점 \space (Weighted \space Rating) = (v/(v+m)) * R + (m/(v+m)) * C $$

- $v$: 개별 영화에 평점을 투표한 횟수
- $m$: 평점을 부여하기 위한 최소 투표 횟수
- $R$: 개별 영화에 대한 평균 평점
- $C$: 전체 영화에 대한 평균 평점

In [37]:
C = movies_df['vote_average'].mean()
m = movies_df['vote_count'].quantile(0.6) #상위 60%에 해당하는 횟수를 기준으로 함
print('C:', round(C,3), 'm:', round(m,3))

C: 6.092 m: 370.2


- `weighted_voet_average()`는 DataFrame의 레코드를 인자로 받아 이 레코드의 vote_count와 vote_averate 칼럼, 그리고 미리 추출된 m과 C 값을 적용해 레코드별 가중 평점을 반환함

In [38]:
percentile = 0.6
m = movies_df['vote_count'].quantile(percentile)
C = movies_df['vote_average'].mean()

def weighted_vote_average(record):
    v = record['vote_count']
    R = record['vote_average']

    return ( (v/(v+m) * R) + (m/(m+v) * C) )

movies_df['weighted_vote'] = movies.apply(weighted_vote_average, axis=1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()


In [39]:
movies_df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                          genres  vote_average  vote_count  \
0  [Action, Adventure, Fantasy, Science Fiction]           7.2       11800   
1                   [Adventure, Fantasy, Action]           6.9        4500   
2                     [Action, Adventure, Crime]           6.3        4466   
3               [Action, Crime, Drama, Thriller]           7.6        9106   
4           [Action, Adventure, Science Fiction]           6.1        2124   

   popularity  \
0  150.437577   
1  139.082615   
2  107.376788   
3  112.312950   
4   43.926995   

                                                                                              keywords  \
0  [culture clash, future, space war, space colony, society, space travel, futuristic, romance, spa...   
1  [ocean, drug abuse, exotic island, east india trading company, love of one's life, traitor, ship...   
2             [spy, based on novel, secret agent, sequel, mi6, british secret service, united kingdom]   
3  [dc comics, crime fighter, terrorist, secret identity, burglar, hostage drama, time bomb, gotham...   
4  [based on novel, mars, medallion, space travel, princess, alien, steampunk, martian, escape, edg...   

                                                                                              overview  \
0  In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, ...   
1  Captain Barbossa, long believed to be dead, has come back to life and is headed to the edge of t...   
2  A cryptic message from Bond’s past sends him on a trail to uncover a sinister organization. Whil...   
3  Following the death of District Attorney Harvey Dent, Batman assumes responsibility for Dent's c...   
4  John Carter is a war-weary, former military captain who's inexplicably transported to the myster...   

                              genre_literal  weighted_vote  
0  Action Adventure Fantasy Science Fiction       7.166301  
1                  Adventure Fantasy Action       6.838594  
2                    Action Adventure Crime       6.284091  
3               Action Crime Drama Thriller       7.541095  
4          Action Adventure Science Fiction       6.098838

In [40]:
movies_df[['title','vote_average','weighted_vote','vote_count']].sort_values('weighted_vote', ascending=False)[:10]

title  vote_average  weighted_vote  vote_count
1881  The Shawshank Redemption           8.5       8.396052        8205
3337             The Godfather           8.4       8.263591        5893
662                 Fight Club           8.3       8.216455        9413
3232              Pulp Fiction           8.3       8.207102        8428
65             The Dark Knight           8.2       8.136930       12002
1818          Schindler's List           8.3       8.126069        4329
3865                  Whiplash           8.3       8.123248        4254
809               Forrest Gump           8.2       8.105954        7927
2294             Spirited Away           8.3       8.105867        3840
2731    The Godfather: Part II           8.3       8.079586        3338

In [42]:
def find_sim_movie(df, sorted_ind, title_name, top_n=10):
    title_movie = df[df['title'] == title_name]
    title_index = title_movie.index.values

    #top_n의 2배에 해당하는 장르 유사성이 높은 인덱스 추출
    similar_indexes = sorted_ind[title_index, :(top_n*2)]
    similar_indexes = similar_indexes.reshape(-1)

    #기준 영화 인덱스는 제외
    similar_indexes = similar_indexes[similar_indexes != title_index]

    #top_n의 2배에 해당하는 후보군에서 weighted_vote가 높은 순으로 top_n만큼 추출    
    return df.iloc[similar_indexes].sort_values('weighted_vote', ascending=False)[:top_n]

similar_movies = find_sim_movie(df=movies_df, sorted_ind=genre_sim_sorted_ind, title_name='The Godfather', top_n=10)
similar_movies[['title','vote_average','weighted_vote']]

title  vote_average  weighted_vote
2731       The Godfather: Part II           8.3       8.079586
1847                   GoodFellas           8.2       7.976937
3866                  City of God           8.1       7.759693
1663  Once Upon a Time in America           8.2       7.657811
883           Catch Me If You Can           7.7       7.557097
281             American Gangster           7.4       7.141396
4041              This Is England           7.4       6.739664
1149              American Hustle           6.8       6.717525
1243                 Mean Streets           7.2       6.626569
2839                     Rounders           6.9       6.530427